samza-sql-shell/src/main/java/org/apache/samza/sql/client/cli/CliEnvironment.java

```diff
@@ -19,112 +19,168 @@
-    System.setErr(stderr);
+  public String[] getPossibleValues(String envName)
+  {
+    EnvironmentVariableHandler handler = getAppropriateHandler(envName);
+    if(handler == null)
+      return null;
+
+    EnvironmentVariableSpecs.Spec spec = handler.getSpecs().getSpec(envName);
+    return spec == null ? null : spec.getPossibleValues();
   }
 
-  void takeEffect() {
-    if (debug) {
-      enableJavaSystemOutAndErr();
-    } else {
-      // We control terminal directly; Forbid any Java System.out and System.err stuff so
-      // any underlying output will not mess up the console
-      disableJavaSystemOutAndErr();
+  public void printAll(PrintWriter writer) {
+    printVariable(writer, CliConstants.CONFIG_EXECUTOR, activeExecutorClassName);
+    printAll(writer, shellEnvHandler);
+    if(executorEnvHandler != null) {
+      printAll(writer, executorEnvHandler);
     }
   }
 
-  private class NullOutputStream extends OutputStream {
-    public void close() {
+  /**
+   * Gives CliEnvironment a chance to apply settings, especially during initialization, things like
+   * making default values take effect
+   */
+  public void finishInitialization() {
+    if(executor == null) {
+      if(createShellExecutor(CliConstants.DEFAULT_EXECUTOR_CLASS)) {
+        activeExecutorClassName = CliConstants.DEFAULT_EXECUTOR_CLASS;
+        executorEnvHandler = executor.getEnvironmentVariableHandler();
+        if(delayedExecutorVars != null) {
+          for (Map.Entry<String, String> entry : delayedExecutorVars.entrySet()) {
+            setEnvironmentVariable(entry.getKey(), entry.getValue());
+          }
+          delayedExecutorVars = null;
+        }
+      } else {
+        throw new CliException("Failed to create default executor: " + CliConstants.DEFAULT_EXECUTOR_CLASS);
+      }
     }
 
-    public void flush() {
+    finishInitialization(shellEnvHandler);
+    finishInitialization(executorEnvHandler);
+  }
+
+  private void finishInitialization(EnvironmentVariableHandler handler) {
+    List<Pair<String, EnvironmentVariableSpecs.Spec>> list = handler.getSpecs().getAllSpecs();
+    for(Pair<String, EnvironmentVariableSpecs.Spec> pair : list) {
+      String envName = pair.getL();
+      EnvironmentVariableSpecs.Spec spec = pair.getR();
+      if(CliUtil.isNullOrEmpty(handler.getEnvironmentVariable(envName))) {
+        handler.setEnvironmentVariable(envName, spec.getDefaultValue());
+      }
     }
+  }
+
+  public SqlExecutor getExecutor() {
+    return executor;
+  }
 
-    public void write(byte[] b) {
+  private boolean createShellExecutor(String executorClassName) {
+    try {
+      Class<?> clazz = Class.forName(executorClassName);
+      Constructor<?> ctor = clazz.getConstructor();
+      executor = (SqlExecutor) ctor.newInstance();
+    } catch (ClassNotFoundException | NoSuchMethodException
+            | IllegalAccessException | InstantiationException | InvocationTargetException e) {
+      return false;
```

Comment:
Whenever an exception is swallowed, We should log it otherwise it will be hard to debug. Can you please log the exception that is swallowed here?

Reply:
Of course. Good point.